### DISCUSSION TITLE
Tabs in custom page

### DISCUSSION THREAD

**Discussion by Lmntrixo** [filamentphp/filament#10127] [Help]
hello there! please need a help on how to make a tabs on custompage in filament?
i have create a custom page named product linked with a resource named categoryResource.Then when i tried to use the getTabs method on this page in show nothing on the panel. 

i have notice that my custom page extends page but not Listrecords, also is it possible to join it to livewire component as it is done whith the table? i mean in the blade file whe use {{$this->table}} for table, is it not possible for tabs too?

here is the code of the product page below

<?php

namespace App\Filament\Resources\CategorieResource\Pages;

use App\Filament\Resources\CategorieResource;
use Filament\Resources\Pages\Page;
use Filament\Resources\Tables\Columns;
use Filament\Resources\Tables\Actions;

use Filament\Tables\Concerns\InteractsWithTable;
use Filament\Tables\Table;
use App\Models\Categorie;
use App\Models\Produit ;
use Filament\Forms\Components\Select;

use Filament\Tables\Columns\TextColumn;
use Filament\Tables\Contracts\HasTable;
use Filament\Resources\Components\Tab;
use Illuminate\Contracts\Routing\UrlRoutable;

class product extends Page implements HasTable
{

    use  InteractsWithTable ;
    protected static string $resource = CategorieResource::class;

    protected static string $view = 'filament.resources.categorie-resource.pages.product';
    

    public Categorie $record;

    public function getTableQuery()
    {
        return Produit::query()->where('idcategorie', $this->record->id);
    }

    public function getTabs(): array
{
    return [
        'all'=>Tab::make('nnnnnnn'),//this method does not work
    ];
}

    

   protected function getTableColumns(): array
    {
        return [
          TextColumn::make('libelle_produit')->label('Produit'),
          TextColumn::make('reference')->label('REFERENCE'),
           
        ];
    }

    public function getTitle(): string 
{
    return $this->record->libelle_categorie;;
}

    
}

**Comment by cntabana** [comment]
You can use this trait **HasTabs**.
<img width="848" alt="Screenshot 2023-12-14 at 15 28 29" src="https://github.com/filamentphp/filament/assets/6320126/785d8425-06fd-4250-8f70-9117cf5c9d0d">


Then add this line on your view: **<x-filament-panels::resources.tabs />**
<img width="536" alt="Screenshot 2023-12-14 at 15 30 34" src="https://github.com/filamentphp/filament/assets/6320126/26394c0e-605e-4531-8e1a-27714646eda1">


Then your **getTabs()** will appear above your table.

**Reply by anindyoy** [reply]
I was following this, but sadly, in filament v4 it won't work anymore, with error message
Unable to locate a class or view for component [filament-panels::resources.tabs].

Did anyone know how to make it work again?

**Reply by krypton612** [reply]
Hi, this video explain showing how to config filamentPHP 4 custom page with tabs table [Video Youtube](https://www.youtube.com/watch?v=x0KD7rstFjc) on your blog explain details [Deatils Explain](https://www.krypton612.ru.eu.org/blog/filament-v4-custom-table-guide)

<img width="1158" height="484" alt="filament-v4-custom-table-guide-result" src="https://github.com/user-attachments/assets/51b5a5f5-4949-41c5-86c5-b8a0795842bb" />

**Comment by photojo** [comment]
How can I solve ...
<x-filament-panels::resources.tabs /> 

throws an error -> getRenderHookScopes does not exist.


I'm using Filament 3, but just tables and forms, not the panel.
Is ist possible to use tabs to filter the table? How get I the tabs above the table.

Thank you very much!

**Reply by yuvrajtimalsina** [reply]
Any progress on this?

**Comment by photojo** [comment]
Unfortunately not.

**Comment by cntabana** [comment]
Can you show me your Livewire page

**Reply by yuvrajtimalsina** [reply]
Just followed the documentation and added what you mentioned in previous response.


https://filamentphp.com/docs/3.x/tables/adding-a-table-to-a-livewire-component

```
<div>

<x-filament-panels::resources.tabs/>

   {{ $this->table }}

</div>
```

The error is :

getRenderHookScopes does not exist.

**Reply by photojo** [reply]
I use the same code and get exactly the same error.

**Comment by cntabana** [comment]
### Livewire Page:

`<?php

namespace App\Filament\Admin\Pages;

use Filament\Forms;
use App\Models\Post;
use Filament\Tables;
use Filament\Pages\Page;
use Filament\Tables\Table;
use Illuminate\Database\Eloquent\Model;
use Filament\Resources\Concerns\HasTabs;
use Illuminate\Database\Eloquent\Builder;
use Filament\Resources\Pages\ListRecords\Tab;

class Posts extends Page implements Tables\Contracts\HasTable, Forms\Contracts\HasForms
{
    use Forms\Concerns\InteractsWithForms;

    use HasTabs;
    use Tables\Concerns\InteractsWithTable {
        makeTable as makeBaseTable;
    }

    protected static ?string $navigationIcon = 'heroicon-o-document-text';

    protected static string $view = 'filament.admin.pages.posts';

    protected function getTableQuery(): Builder
    {
        return Post::query();
    }

    public function table(Table $table): Table
    {
        return $table
            ->query($this->getTableQuery())
            ->columns([
                    Tables\Columns\TextColumn::make('title'),
            ]);
    }


    public function getTabs(): array
    {
        return [

            'publish' => Tab::make('Published')
                        ->icon('heroicon-m-user')
                        ->badge(fn () => Post::where('status', true)->count())
                        ->modifyQueryUsing(fn (Builder $query) => $query->where('status', true)),

            'un_publish' => Tab::make('UnPublished')
                        ->icon('heroicon-m-user')
                        ->badge(fn () => Post::where('status', false)->count())
                        ->modifyQueryUsing(fn (Builder $query) => $query->where('status', false)),
        ];
    }

    protected function makeTable(): Table
    {
        return $this->makeBaseTable()
            ->query(fn (): Builder => $this->getTableQuery())
            ->modifyQueryUsing($this->modifyQueryWithActiveTab(...))
            //->modelLabel($this->getModelLabel() ?? static::getResource()::getModelLabel())
            //->pluralModelLabel($this->getPluralModelLabel() ?? static::getResource()::getPluralModelLabel())
            ->recordAction(function (Model $record, Table $table): ?string {
                foreach (['view', 'edit'] as $action) {
                    $action = $table->getAction($action);

                    if (! $action) {
                        continue;
                    }

                    $action->record($record);

                    if ($action->isHidden()) {
                        continue;
                    }

                    if ($action->getUrl()) {
                        continue;
                    }

                    return $action->getName();
                }

                return null;
            })
            //->recordTitle(fn (Model $record): string => static::getResource()::getRecordTitle($record))
            ->recordUrl($this->getTableRecordUrlUsing() ?? function (Model $record, Table $table): ?string {
                foreach (['view', 'edit'] as $action) {
                    $action = $table->getAction($action);

                    if (! $action) {
                        continue;
                    }

                    $action->record($record);

                    if ($action->isHidden()) {
                        continue;
                    }

                    $url = $action->getUrl();

                    if (! $url) {
                        continue;
                    }

                    return $url;
                }
                return null;
            });

    }
}
`

### View Page:

`<x-filament::page>
    <x-filament-panels::resources.tabs />
        {{ $this->table }}
</x-filament::page>
`

I have tested this one and it works.

**Comment by yuvrajtimalsina** [comment]
**DonorList.php**

```
<?php

namespace App\Livewire;

use App\Models\Donor;
use Filament\Forms\Concerns\InteractsWithForms;
use Filament\Forms\Contracts\HasForms;
use Filament\Resources\Components\Tab;
use Filament\Resources\Concerns\HasTabs;
use Filament\Tables\Columns\TextColumn;
use Filament\Tables\Concerns\InteractsWithTable;
use Filament\Tables\Contracts\HasTable;
use Filament\Tables\Table;
use Illuminate\Contracts\View\View;
use Illuminate\Database\Query\Builder;
use Livewire\Component;

class DonorsList extends Component implements HasForms, HasTable
{
    use HasTabs;
    use InteractsWithForms;
    use InteractsWithTable;

    public function getTabs(): array
    {
        return [

            'O+' => Tab::make('O+')
                ->icon('heroicon-m-user')
                ->modifyQueryUsing(fn (Builder $query) => $query->where('blood_group', 'O+')),
        ];
    }

    public function table(Table $table): Table
    {
        return $table
            ->query(Donor::query())
            ->columns([
                TextColumn::make('name'),
            ]);
    }

    public function render(): View
    {
        return view('livewire.donors-list');
    }
}
```
**donors-list.blade.php**
```
<div>
    <x-filament-panels::resources.tabs/>
    {{ $this->table }}
</div>
```
**Error**
![image](https://github.com/filamentphp/filament/assets/107990171/f75b819c-6f19-4ded-a2fb-ea4be6b7e5ed)

**Comment by cntabana** [comment]
Try to add this function in your class 

`public function getRenderHookScopes(): array 
    {
        return [static::class];
    }`

**Reply by yuvrajtimalsina** [reply]
It seems to display the tabs even if the array is empty but data are not filtered.

**Comment by photojo** [comment]
I have a similar code, but now I get
_Method App\Livewire\Model::getCachedTabs does not exist._

**Comment by cntabana** [comment]
use this trait like this 

`use Tables\Concerns\InteractsWithTable {
    makeTable as makeBaseTable;
}`

and add this function in yr class
`
protected function makeTable(): Table
{
    return $this->makeBaseTable()
        ->query(fn (): Builder => $this->getTableQuery())
        ->modifyQueryUsing($this->modifyQueryWithActiveTab(...))
        //->modelLabel($this->getModelLabel() ?? static::getResource()::getModelLabel())
        //->pluralModelLabel($this->getPluralModelLabel() ?? static::getResource()::getPluralModelLabel())
        ->recordAction(function (Model $record, Table $table): ?string {
            foreach (['view', 'edit'] as $action) {
                $action = $table->getAction($action);

                if (! $action) {
                    continue;
                }

                $action->record($record);

                if ($action->isHidden()) {
                    continue;
                }

                if ($action->getUrl()) {
                    continue;
                }

                return $action->getName();
            }

            return null;
        })
        //->recordTitle(fn (Model $record): string => static::getResource()::getRecordTitle($record))
        ->recordUrl($this->getTableRecordUrlUsing() ?? function (Model $record, Table $table): ?string {
            foreach (['view', 'edit'] as $action) {
                $action = $table->getAction($action);

                if (! $action) {
                    continue;
                }

                $action->record($record);

                if ($action->isHidden()) {
                    continue;
                }

                $url = $action->getUrl();

                if (! $url) {
                    continue;
                }

                return $url;
            }
            return null;
        });

}
`

**Comment by cntabana** [comment]
@photojo 

run `php artisan optimize:clear`

**Comment by photojo** [comment]
Same problem.

When I use  @dump($this->getTabs()) in my blade-file, I get all tabs.

But using <x-filament-panels::resources.tabs /> I still get 
_getCachedTabs does not exist_

**Comment by cntabana** [comment]
try this 
`public function getCachedTabs(): array
    {
        return $this->getTabs();
    }`

ref: https://discord.com/channels/883083792112300104/1181239883965157467/1181307522389377066

**Reply by yuvrajtimalsina** [reply]
@cntabana Tabs seem to work fine after using the trait but It should work just by using the trait instead I had to duplicate code fragment in both places. Can you explain me the logic behind it?

**Comment by cntabana** [comment]
The tab is a feature for resources, so to make it work on the customer page, we need to utilize the function located under Filament\Resources\Pages\ListRecords (**protected function makeTable(): Table**). I am confident that there is an easy way to achieve this without duplicating the code, but currently, I don't have enough time to figure it out. However, later this week, I will dedicate time to find the solution of this. If I find the solution, I will post it here

**Comment by photojo** [comment]
Would it be possible to describe wich files you are using at the moment? I'm a little bit confused now. I'm so sorry.

**Reply by yuvrajtimalsina** [reply]
**Make trait as mentioned here :**

_Livewire\Tables\Concerns\InteractsWithTable.php_

https://github.com/filamentphp/filament/discussions/10127#discussioncomment-8208409

**Take reference from code at the beginning :**

https://github.com/filamentphp/filament/discussions/10127#discussioncomment-8206452

**Reply by photojo** [reply]
Tank you very much!

But I get '\Tab::modifyQuery does not exist' when clicking on one of the tabs?

**Reply by cntabana** [reply]
Share yr code please

**Comment by photojo** [comment]
My component:

`<?php

namespace App\Livewire;

use Livewire\Component;
use App\Models\Teacher;
use Filament\Tables\Table;
use Filament\Forms\Contracts\HasForms;
use Filament\Forms\Components\Tabs\Tab;
use Filament\Tables\Columns\TextColumn;
use Filament\Tables\Contracts\HasTable;
use Illuminate\Database\Eloquent\Builder;
use Filament\Forms\Concerns\InteractsWithForms;
use Filament\Tables\Concerns\InteractsWithTable;
use Filament\Resources\Concerns\HasTabs;

class Registrations extends Component implements HasForms, HasTable
{
    use HasTabs;
    use InteractsWithTable;
    use InteractsWithForms;


    use HasTabs;
    use InteractsWithTable {
        makeTable as makeBaseTable;
    }


    public function table(Table $table): Table
    {
        return $table
            ->query(
                Teacher::query()

            )
            ->columns([
                TextColumn::make('status'),
            ])

            ->paginated([10, 'all']);
    }


    public function getTabs(): array
    {
        return [
            'all' => Tab::make(('All')),
         

            'new' => Tab::make('new')
                ->icon('heroicon-m-user')
                ->badge(fn () => Teacher::where('status', 'new')->count())
                // ->modifyQueryUsing(fn (Builder $query) => $query->where('status', 'new'))


        ];
    }

    protected function makeTable(): Table
    {
        return $this->makeBaseTable()
            ->query(fn (): Builder => $this->getTableQuery())
            ->modifyQueryUsing($this->modifyQueryWithActiveTab(...))
            ->recordAction(function (Teacher $record, Table $table): ?string {
                foreach (['view', 'edit'] as $action) {
                    $action = $table->getAction($action);

                    if (!$action) {
                        continue;
                    }

                    $action->record($record);

                    if ($action->isHidden()) {
                        continue;
                    }

                    if ($action->getUrl()) {
                        continue;
                    }

                    return $action->getName();
                }

                return null;
            });
    }

    public function getCachedTabs(): array
    {
        return $this->getTabs();
    }

    public function getRenderHookScopes(): array
    {
        return [static::class];
    }

}
`

My blade-file

________________________
'<div>

    <x-filament-panels::resources.tabs />
    {{ $this->table }}
'</div>

________________________

And my Trait/Consern

`<?php

use App\Models\Teacher;
use Filament\Tables\Table;
use Illuminate\Database\Eloquent\Builder;

trait InteractsWithTable
{


    protected function makeTable(): Table
    {
        return $this->makeBaseTable()
            ->query(fn (): Builder => $this->getTableQuery())
            ->modifyQueryUsing($this->modifyQueryWithActiveTab(...))
            //->modelLabel($this->getModelLabel() ?? static::getResource()::getModelLabel())
            //->pluralModelLabel($this->getPluralModelLabel() ?? static::getResource()::getPluralModelLabel())
            ->recordAction(function (Teacher $record, Table $table): ?string {
                foreach (['view', 'edit'] as $action) {
                    $action = $table->getAction($action);

                    if (!$action) {
                        continue;
                    }

                    $action->record($record);

                    if ($action->isHidden()) {
                        continue;
                    }

                    if ($action->getUrl()) {
                        continue;
                    }

                    return $action->getName();
                }

                return null;
            })
            //->recordTitle(fn (Model $record): string => static::getResource()::getRecordTitle($record))
            ->recordUrl($this->getTableRecordUrlUsing() ?? function (Teacher $record, Table $table): ?string {
                foreach (['view', 'edit'] as $action) {
                    $action = $table->getAction($action);

                    if (!$action) {
                        continue;
                    }

                    $action->record($record);

                    if ($action->isHidden()) {
                        continue;
                    }

                    $url = $action->getUrl();

                    if (!$url) {
                        continue;
                    }

                    return $url;
                }
                return null;
            });
    }
}
`

**Comment by cntabana** [comment]
I tried to change your component like this:

`
namespace App\Livewire;

use Livewire\Component;
use App\Models\Teacher;
use Filament\Tables\Table;
use Filament\Forms\Contracts\HasForms;
use Filament\Forms\Components\Tabs\Tab;
use Filament\Tables\Columns\TextColumn;
use Filament\Tables\Contracts\HasTable;
use Illuminate\Database\Eloquent\Builder;
use Filament\Forms\Concerns\InteractsWithForms;
use Filament\Tables\Concerns\InteractsWithTable;
use Filament\Resources\Concerns\HasTabs;

class Registrations extends Component implements HasForms, HasTable
{

    use HasTabs;
    use InteractsWithForms;
    use InteractsWithTable {
        makeTable as makeBaseTable;
    }

    public function table(Table $table): Table
    {
        return $table
            ->query(
                Teacher::query()

            )
            ->columns([
                TextColumn::make('status'),
            ])

            ->paginated([10, 'all']);
    }


    public function getTabs(): array
    {
        return [
            'all' => Tab::make(('All')),
         

            'new' => Tab::make('new')
                ->icon('heroicon-m-user')
                ->badge(fn () => Teacher::where('status', 'new')->count())
                // ->modifyQueryUsing(fn (Builder $query) => $query->where('status', 'new'))


        ];
    }

    protected function makeTable(): Table
    {
        return $this->makeBaseTable()
            ->modifyQueryUsing($this->modifyQueryWithActiveTab(...))
            ->recordAction(function (Teacher $record, Table $table): ?string {
                foreach (['view', 'edit'] as $action) {
                    $action = $table->getAction($action);

                    if (!$action) {
                        continue;
                    }

                    $action->record($record);

                    if ($action->isHidden()) {
                        continue;
                    }

                    if ($action->getUrl()) {
                        continue;
                    }

                    return $action->getName();
                }

                return null;
            });
    }

    public function getCachedTabs(): array
    {
        return $this->getTabs();
    }

    public function getRenderHookScopes(): array
    {
        return [static::class];
    }
}
`

If it doesn't work, please create a repository, and I will take a look for you.

**Comment by cntabana** [comment]
@yuvraj-timalsina 

You can create a trait looks like this 
'<?php
namespace App\Filament\Traits\Filament;

use Filament\Tables;
use Filament\Tables\Table;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;

trait InteractWithTabs
{

    use Tables\Concerns\InteractsWithTable {
        makeTable as makeBaseTable;
    }

    protected function makeTable(): Table
    {
        return $this->makeBaseTable()
            ->query(fn (): Builder => $this->getTableQuery())
            ->modifyQueryUsing($this->modifyQueryWithActiveTab(...));
    }
}
`

Then use that trait (use InteractWithTabs) anywhere you want.

**Reply by yuvrajtimalsina** [reply]
@cntabana This is neat and works like a charm. Thank you for the support. Happy Coding!

**Comment by photojo** [comment]
Thank you everyone how wanted to help me. Now, after several hours I give up and use the  usual filter-functionality.
Greetings 
Jo

**Reply by yuvrajtimalsina** [reply]
You may still share your code once. It should work.

**Reply by photojo** [reply]
@yuvraj-timalsina 
Here you can find my code: [https://github.com/photojo/laravel3_tabs_filter](https://github.com/photojo/laravel3_tabs_filter)

Thank you very much!

**Reply by yuvrajtimalsina** [reply]
**InteractWithTabsTrait**
```
<?php

namespace App\Traits;

use Filament\Tables\Concerns\InteractsWithTable;
use Filament\Tables\Table;
use Illuminate\Database\Eloquent\Builder;

trait InteractWithTabsTrait
{
    use InteractsWithTable {makeTable as makeBaseTable; }

    protected function makeTable(): Table
    {
        return $this->makeBaseTable()
            ->query(fn (): Builder => $this->getTableQuery())
            ->modifyQueryUsing($this->modifyQueryWithActiveTab(...));
    }
}
```
**TeacherComponent**

```
<?php

namespace App\Livewire;

use App\Models\Teacher;
use App\Traits\InteractWithTabsTrait;
use Filament\Forms\Concerns\InteractsWithForms;
use Filament\Forms\Contracts\HasForms;
use Filament\Pages\Page;
use Filament\Resources\Concerns\HasTabs;
use Filament\Resources\Pages\ListRecords\Tab;
use Filament\Tables;
use Filament\Tables\Contracts\HasTable;
use Filament\Tables\Table;
use Illuminate\Database\Eloquent\Builder;

class TeacherComponent extends Page implements HasForms, HasTable
{
    use HasTabs;
    use InteractsWithForms;
    use InteractWithTabsTrait;

    protected static string $view = 'livewire.teacher-component';

    public function getTabs(): array
    {
        $tabs = ['All' => Tab::make()];
        $teacherStatuses = Teacher::whereNotNull('status')
            ->pluck('status')
            ->unique();

        foreach ($teacherStatuses as $teacherStatus) {
            $tabs[$teacherStatus] = Tab::make()
                ->modifyQueryUsing(function (Builder $query) use ($teacherStatus) {
                    return $query->where('status', $teacherStatus);
                });
        }

        return $tabs;
    }

    public function table(Table $table): Table
    {
        return $table
            ->query(Teacher::query())
            ->columns([
                Tables\Columns\TextColumn::make('name')
                    ->searchable(),
                Tables\Columns\TextColumn::make('status')
                    ->searchable(),
            ]);
    }
}
```

**Note**: _Do not use full-page Livewire Component for the table, instead render it inside your other component or page._

```
<livewire:teacher-component/>
```

**Reply by photojo** [reply]
Oh man! It works! Thank you so much!
Just two additional Questions:

1. Note: Do not use full-page Livewire Component for the table, instead render it inside your other component or page. Why?
2. So, tabs for filter tables do only work with this trait, not out of the "filament-box"?

**Reply by yuvrajtimalsina** [reply]
Firstly, it only worked by extending the Page not the Component and if used with full-page Livewire Component gives off dashboard navigation and navbar as well so I recommended the way it works.

Secondly, it only worked by utilizing the trait for now so be it, maybe it will work out of the box in later releases.

Happy Coding!

**Comment by photojo** [comment]
Ok! All right!
Tanks a lot!

**Comment by ShaungBhone** [comment]
It's not working for me. I got getAttribute not defined in javascript browser console.

**Reply by yuvrajtimalsina** [reply]
Did you follow the instructions mentioned?
 
https://github.com/filamentphp/filament/discussions/10127#discussioncomment-8212403

**Reply by ShaungBhone** [reply]
Yes

**Reply by yuvrajtimalsina** [reply]
Could you share your code?

**Reply by ShaungBhone** [reply]
Currently I skip this step. I will try later.

**Comment by KhairulAzmi21** [comment]
Can we use tabs inside livewire component ? not inside filament pages.

**Comment by rzortiz11** [comment]
Cool it works, but how can we set it to have a default selected tabs?


This seems not working 

    public function getDefaultActiveTab(): string | int | null
    {
        // not working
        return 'June';
    }

**Reply by yuvrajtimalsina** [reply]
@ZerrudOrtiz 

Try the following method :
 https://filamentphp.com/docs/3.x/infolists/layout/tabs#setting-the-default-active-tab
```
->activeTab(2)
```

**Reply by rzortiz11** [reply]
Hi @yuvraj-timalsina thanks for the response just to confirm will it work on this? the tab is above the table header, like what the discussion is about.

![image](https://github.com/filamentphp/filament/assets/156640289/50f15589-25df-453e-b9ff-e2c17be53dbf)

**Reply by yuvrajtimalsina** [reply]
Not sure, you may try with int value as well.Else, if it didn't work, feel free to share code and we will work something out together.

**Reply by shaheenfawzy** [reply]
You need to call $this->loadDefaultActiveTab() in the mount method inside the page like this:
```php
public function mount(): void
{
    $this->loadDefaultActiveTab();
}

public function getDefaultActiveTab(): string|int|null
{
    return 'June';
}
```
